### PR TITLE
List circularizer

### PR DESCRIPTION
**Pull request by JornVernee**
Seeing https://github.com/CleverRaven/Cataclysm-DDA/pull/14192 reminded me of some other for loops that could do with some touch up (in `crafting_gui.cpp`).

The problem was that they were bulky, partly because they tried to access a list as if it were circular.

So to 'clean up' the for loops and to solve the general problem of accessing a list as if it were circular, I made `list_circularizer` which wraps a list so it gives the impression of looping around on itself. I then exchanged it with the functions containing the for loops.

The idea is that this wrapper only reads from a list, so it stores a pointer to the list. That means that the list has to exist separately. But that also means that changes to the list propagate both ways (and it reduces overhead). Some feedback on that idea would be nice.

### CODE REVIEW

File: src/cata_utility.h [hunk 1]
```diff
@@ -42,4 +42,48 @@ double logarithmic_range( int min, int max, int pos );
...
+* This wrapper class allows you to do that. It requires the list to exist
+* separately, but that also means any changes to the list get propagated (both ways).
+*/
+template<typename T>
+class list_circularizer {
+    private:
+        unsigned int index = 0;
+        std::vector<T> *_list;
+    public:
+        list_circularizer( std::vector<T> &_list ) : _list( &_list )
```
Comment (kevingranade) [inline src/cata_utility.h]: Small nit, we use 1tbs, so this would be:
```
list_circularizer( std::vector<T> &_list ) : _list( &_list ) {
}
```

Comment (JornVernee) [inline src/cata_utility.h]: On all function headers? Or just on this one? ```CODE_STYLE.txt``` has them on the next line. (I personally prefer the style in your correction though).

Comment (kevingranade) [inline src/cata_utility.h]: Generally it is all blocks that are not at the top level.  So in-line
method definitions like this put brackets on the same line, but method
definitions in cop files will put them on the next line.

**Comment by BevapDin** [comment]
I like the idea, just some nitpicking:
- the member names are inconsistent: `index` and `_list` (why has the later an underscore?)

- `operator()` for changing the list seems strange, I would have assumed `operator()` returns the current element (similar to `operator[]`, just without the parameter). Setting a different list could be done via an assignment operator. But this part is very subjective.

For the future:
- The `next`/`prev` function may need a parameter (or overloaded versions that have a parameter) that allow scrolling a whole page (supplying the size of a page is the callers responsibility).

**Comment by JornVernee** [comment]
```list``` gets highlighted by C::B, because it's STL. I didn't want to have that illusion so I changed it. In hindsight there was the inconsistency, but I didn't have time to change the name of ```index``` yet.

I mainly used ```operator()``` to change the list because ```x = list_circularizer<std::string>( some_list )``` is rather long, and the operator was still unused. I think that something like ```list_circ( my_list )``` is clear enough. You're putting a new list into a ```list_circularizer```, and with the context of what a ```list_circularizer``` is (which can be obtained through the documentation in the code), it should be clear enough that ```list_circ``` is now reading from the new list.

As for the page scrolling, let's remember that this is a template class, and like you said, it's the caller responsibility to supply the size of a page. So why not supply a ```page``` as a template argument:
```c++
std::vector<page> pages {...}; // Where page is a user defined type, or just an std::vector<std::string>
list_circularizer<page> circ_pages( pages );
```
Adding page scrolling functionality feels like defying SRP, since we cater to multiple kinds of clients (ones that want to scroll through just a list, and ones that want to scroll through a list *and* through pages).

**Comment by BevapDin** [comment]
> list  gets highlighted by C::B, because it's STL. 

`std::list` is STL. `list` itself (in a non-global scope) is absolutely fine.

Regarding the page: the caller only needs to supply the page size, not a `page` thing, it's about scrolling N steps (instead of 1) at once.

**Comment by JornVernee** [comment]
I know ```list``` is not strictly STL, but it still gets highlighted in C::B, which I didn't want.

I know what you meant about scrolling a page (scrolling n steps at a time instead of just 1), I was just suggesting a different solution that I consider to be cleaner (from an OO perspective). The idea is that when you scroll a page you use ```circ_list.next()``` and when scrolling a line use ```circ_list.cur().next()``` (where ```circ_list``` 'contains' another ```list_circularizer<...>```, or some other class that suits the situation).

**Comment by kevingranade** [comment]
The only utilized functionality is the constructor, prev(), next(), and cur(), why even have the overloads?  YAGNI

**Pull request merged**